In [2]:
import numpy as np 
import pandas as pd
import matplotlib as plt

In [3]:
df = pd.read_csv("startups.csv")

In [4]:
df.head(10)

name     city  \
0          Campus Bubble  Atlanta   
1               DueProps  Atlanta   
2              SalesLoft  Atlanta   
3  The Coca-Cola Company  Atlanta   
4              EarthLink  Atlanta   
5                REscour  Atlanta   
6               viaCycle  Atlanta   
7           Seraph Group  Atlanta   
8                Kabbage  Atlanta   
9          Usable Health  Atlanta   

                                             tagline  \
0                             Your Academic Identity   
1  Gamifying the $46 Billion Employee Incentives ...   
2          Quickly build high-quality prospect lists   
3                                                NaN   
4                                                NaN   
5  Market intelligence and analytics for commerci...   
6  Zipcar for bicycles. Call or text, unlock, and...   
7                                                NaN   
8                                                NaN   
9  Menu-personalization software to attract regulars   

                                         description  
0  Campus Bubble (“CB”) is the Academic Community...  
1                                t unprecedented ...  
2                build high-quality prospect lists\n  
3  Coca-Cola Journey is a digital magazine that f...  
4                                                NaN  
5  REscour is a data platform and decision engine...  
6  viaCycle creates bicycle sharing technology th...  
7                                                NaN  
8  Kabbage delivers small businesses financing. B...  
9  SmartMenus are web-based ordering kiosks and t...

In [5]:
df.isnull().sum()

name              2
city              0
tagline        5307
description    4863
dtype: int64

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 42038 entries, 0 to 42037
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   name         42036 non-null  object
 1   city         42038 non-null  object
 2   tagline      36731 non-null  object
 3   description  37175 non-null  object
dtypes: object(4)
memory usage: 1.3+ MB


We use this dataset to take a start-up's description into consideration. 
As observed from the df.info, we can see that it has 37175 non-null values, compare to a relatively small null values of the description. We can remove the null values

In [7]:
df1 = df.dropna(subset = ['name','description'])

In [8]:
df1.isnull().sum()

name              0
city              0
tagline        1137
description       0
dtype: int64

Now as we have gotten rid of all the null values in the 2 most important columns, we can move on to applying Natural Language Processing algo to deal with text data.

In [9]:
punc = '''!()-[]{};:'"\,<>./?@#$%^&*_~'''

In [10]:
df1['description'] = df1['description'].str.replace('[^\w\s]','')

<ipython-input-10-ba98275d4482>:1: FutureWarning: The default value of regex will change from True to False in a future version.
  df1['description'] = df1['description'].str.replace('[^\w\s]','')
<ipython-input-10-ba98275d4482>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df1['description'] = df1['description'].str.replace('[^\w\s]','')


In [11]:
df1

name           city  \
0                   Campus Bubble        Atlanta   
1                        DueProps        Atlanta   
2                       SalesLoft        Atlanta   
3           The Coca-Cola Company        Atlanta   
5                         REscour        Atlanta   
...                           ...            ...   
42031             Weekend Package  Washington DC   
42032           LogiAnalytics.com  Washington DC   
42035             UP Technologies  Washington DC   
42036  Galaxie restaurant and bar  Washington DC   
42037              Avenger Mobile  Washington DC   

                                                 tagline  \
0                                 Your Academic Identity   
1      Gamifying the $46 Billion Employee Incentives ...   
2              Quickly build high-quality prospect lists   
3                                                    NaN   
5      Market intelligence and analytics for commerci...   
...                                                  ...   
42031                       Worldwide Hotel Deals Online   
42032                                                NaN   
42035       Ultra Portable Commercial Hardware Solutions   
42036            concept of a unique, modern restaurant    
42037  Avenger Mobile aims to create apps that solve ...   

                                             description  
0      Campus Bubble CB is the Academic Community Net...  
1                                       t unprecedented   
2                     build highquality prospect lists\n  
3      CocaCola Journey is a digital magazine that fo...  
5      REscour is a data platform and decision engine...  
...                                                  ...  
42031                                         is modest   
42032  Logi Info embeds interactive visualizations an...  
42035  Ultra Portable Technologies UP Technologies is...  
42036  Concept of a modern 10 15 000 sf restaurant wi...  
42037   Mobile aims to create apps that solve realwor...  

[37173 rows x 4 columns]

In [12]:
import nltk
from nltk.corpus import stopwords

In [13]:
nltk.download('stopwords')
from nltk.tokenize import word_tokenize

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\nhoan\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [14]:
stop = stopwords.words('english')

In [15]:
df1['new_des'] = df1['description'].apply(lambda x: ' '.join([word for word in x.split() if word not in (stop)]))

<ipython-input-15-fdb5a9ef813c>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df1['new_des'] = df1['description'].apply(lambda x: ' '.join([word for word in x.split() if word not in (stop)]))


In [16]:
df1.head(10)

name     city  \
0           Campus Bubble  Atlanta   
1                DueProps  Atlanta   
2               SalesLoft  Atlanta   
3   The Coca-Cola Company  Atlanta   
5                 REscour  Atlanta   
6                viaCycle  Atlanta   
8                 Kabbage  Atlanta   
9           Usable Health  Atlanta   
11              OpenStudy  Atlanta   
12                 We&Co_  Atlanta   

                                              tagline  \
0                              Your Academic Identity   
1   Gamifying the $46 Billion Employee Incentives ...   
2           Quickly build high-quality prospect lists   
3                                                 NaN   
5   Market intelligence and analytics for commerci...   
6   Zipcar for bicycles. Call or text, unlock, and...   
8                                                 NaN   
9   Menu-personalization software to attract regulars   
11                   Social Learning for Open Courses   
12                   People Analytics for Hospitality   

                                          description  \
0   Campus Bubble CB is the Academic Community Net...   
1                                    t unprecedented    
2                  build highquality prospect lists\n   
3   CocaCola Journey is a digital magazine that fo...   
5   REscour is a data platform and decision engine...   
6   viaCycle creates bicycle sharing technology th...   
8   Kabbage delivers small businesses financing Bo...   
9   SmartMenus are webbased ordering kiosks and ta...   
11  rovides realtime study communities for over 40...   
12  WeCo provides People Analytics to the Hospital...   

                                              new_des  
0   Campus Bubble CB Academic Community Network Li...  
1                                       unprecedented  
2                    build highquality prospect lists  
3   CocaCola Journey digital magazine focuses impo...  
5   REscour data platform decision engine utilizes...  
6   viaCycle creates bicycle sharing technology fl...  
8   Kabbage delivers small businesses financing Bo...  
9   SmartMenus webbased ordering kiosks tablets se...  
11  rovides realtime study communities 40 OpenCour...  
12  WeCo provides People Analytics Hospitality ind...

In [17]:
#Make all the words in description lower case
df1['new_des'] = df1['new_des'].str.lower()

<ipython-input-17-4b6a16c8a7e4>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df1['new_des'] = df1['new_des'].str.lower()


In [18]:
df1.head(10)

name     city  \
0           Campus Bubble  Atlanta   
1                DueProps  Atlanta   
2               SalesLoft  Atlanta   
3   The Coca-Cola Company  Atlanta   
5                 REscour  Atlanta   
6                viaCycle  Atlanta   
8                 Kabbage  Atlanta   
9           Usable Health  Atlanta   
11              OpenStudy  Atlanta   
12                 We&Co_  Atlanta   

                                              tagline  \
0                              Your Academic Identity   
1   Gamifying the $46 Billion Employee Incentives ...   
2           Quickly build high-quality prospect lists   
3                                                 NaN   
5   Market intelligence and analytics for commerci...   
6   Zipcar for bicycles. Call or text, unlock, and...   
8                                                 NaN   
9   Menu-personalization software to attract regulars   
11                   Social Learning for Open Courses   
12                   People Analytics for Hospitality   

                                          description  \
0   Campus Bubble CB is the Academic Community Net...   
1                                    t unprecedented    
2                  build highquality prospect lists\n   
3   CocaCola Journey is a digital magazine that fo...   
5   REscour is a data platform and decision engine...   
6   viaCycle creates bicycle sharing technology th...   
8   Kabbage delivers small businesses financing Bo...   
9   SmartMenus are webbased ordering kiosks and ta...   
11  rovides realtime study communities for over 40...   
12  WeCo provides People Analytics to the Hospital...   

                                              new_des  
0   campus bubble cb academic community network li...  
1                                       unprecedented  
2                    build highquality prospect lists  
3   cocacola journey digital magazine focuses impo...  
5   rescour data platform decision engine utilizes...  
6   viacycle creates bicycle sharing technology fl...  
8   kabbage delivers small businesses financing bo...  
9   smartmenus webbased ordering kiosks tablets se...  
11  rovides realtime study communities 40 opencour...  
12  weco provides people analytics hospitality ind...

# REPEAT THE PROCESS WITH THE INVESTOR DATASET

In [23]:
investor = pd.read_csv('investor_data.csv')
investors = investor.head(20)

In [24]:
investors

Organization Name  \
0           Alumni Ventures Group   
1               Tech Coast Angels   
2   Ontario Centres of Excellence   
3                     StartEngine   
4                     SFC Capital   
5                  Keiretsu Forum   
6                Sand Hill Angels   
7              Alliance of Angels   
8                       VA Angels   
9           Houston Angel Network   
10                       GoBeyond   
11       Keiretsu Forum Northwest   
12                New York Angels   
13                  Mumbai Angels   
14           Indian Angel Network   
15                     Zillionize   
16                    LetsVenture   
17                   Anges Quebec   
18                   VentureSouth   
19           Boston Harbor Angels   

                                          Description  Number of Investments  
0   Alumni Ventures Group provides diversified ven...                    835  
1   Invests in high-growth start-up's headquartere...                    567  
2   The Ontario Centres of Excellence invests in p...                    433  
3   StartEngine is an equity crowdfunding platform...                    349  
4   Early-stage investor combining its own angel s...                    331  
5   Keiretsu Forum is a California-based angel gro...                    291  
6   Sand Hill Angels, a venture capital firm, inve...                    202  
7   We are a group of 140+ active angel investors....                    185  
8   VA Angels, also known as VA Angels, is an Ange...                    184  
9   Houston Angel Network provides capital to earl...                    179  
10  GoBeyond enables individuals to access angel i...                    175  
11  Keiretsu Forum Northwest is a Washington-based...                    172  
12  New York Angels is a professional angel organi...                    167  
13  Mumbai Angels, an Indian angel investment netw...                    163  
14  Indian Angel Network is one of the leading Ang...                    145  
15  Zillionize is a low friction startup investor ...                    145  
16  Let’sVenture enables startups looking to raise...                    137  
17  Anges Québec is a network of 230 angel investo...                    133  
18  VentureSouth is an early stage venture firm th...                    122  
19  Boston Harbor Angels is a group of proven busi...                    120

In [27]:
investors.info() #no null value

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 3 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   Organization Name      20 non-null     object
 1   Description            20 non-null     object
 2   Number of Investments  20 non-null     int64 
dtypes: int64(1), object(2)
memory usage: 608.0+ bytes


In [28]:
punc = '''!()-[]{};:'"\,<>./?@#$%^&*_~'''

In [29]:
investors['Description'] = investors['Description'].str.replace('[^\w\s]','')

<ipython-input-29-854aa1609cb2>:1: FutureWarning: The default value of regex will change from True to False in a future version.
  investors['Description'] = investors['Description'].str.replace('[^\w\s]','')
<ipython-input-29-854aa1609cb2>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  investors['Description'] = investors['Description'].str.replace('[^\w\s]','')


In [30]:
investors['new_des'] = investors['Description'].apply(lambda x: ' '.join([word for word in x.split() if word not in (stop)]))

<ipython-input-30-31750c02bfc5>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  investors['new_des'] = investors['Description'].apply(lambda x: ' '.join([word for word in x.split() if word not in (stop)]))


In [33]:
investors['new_des'] = investors['new_des'].str.lower()

<ipython-input-33-895150a41861>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  investors['new_des'] = investors['new_des'].str.lower()


In [34]:
investors

Organization Name  \
0           Alumni Ventures Group   
1               Tech Coast Angels   
2   Ontario Centres of Excellence   
3                     StartEngine   
4                     SFC Capital   
5                  Keiretsu Forum   
6                Sand Hill Angels   
7              Alliance of Angels   
8                       VA Angels   
9           Houston Angel Network   
10                       GoBeyond   
11       Keiretsu Forum Northwest   
12                New York Angels   
13                  Mumbai Angels   
14           Indian Angel Network   
15                     Zillionize   
16                    LetsVenture   
17                   Anges Quebec   
18                   VentureSouth   
19           Boston Harbor Angels   

                                          Description  Number of Investments  \
0   Alumni Ventures Group provides diversified ven...                    835   
1   Invests in highgrowth startups headquartered i...                    567   
2   The Ontario Centres of Excellence invests in p...                    433   
3   StartEngine is an equity crowdfunding platform...                    349   
4   Earlystage investor combining its own angel sy...                    331   
5   Keiretsu Forum is a Californiabased angel grou...                    291   
6   Sand Hill Angels a venture capital firm invest...                    202   
7   We are a group of 140 active angel investors E...                    185   
8   VA Angels also known as VA Angels is an Angel ...                    184   
9   Houston Angel Network provides capital to earl...                    179   
10  GoBeyond enables individuals to access angel i...                    175   
11  Keiretsu Forum Northwest is a Washingtonbased ...                    172   
12  New York Angels is a professional angel organi...                    167   
13  Mumbai Angels an Indian angel investment netwo...                    163   
14  Indian Angel Network is one of the leading Ang...                    145   
15  Zillionize is a low friction startup investor ...                    145   
16  LetsVenture enables startups looking to raise ...                    137   
17  Anges Québec is a network of 230 angel investo...                    133   
18  VentureSouth is an early stage venture firm th...                    122   
19  Boston Harbor Angels is a group of proven busi...                    120   

                                              new_des  
0   alumni ventures group provides diversified ven...  
1   invests highgrowth startups headquartered sout...  
2   the ontario centres excellence invests project...  
3   startengine equity crowdfunding platform allow...  
4   earlystage investor combining angel syndicate ...  
5   keiretsu forum californiabased angel group off...  
6   sand hill angels venture capital firm invests ...  
7   we group 140 active angel investors each year ...  
8   va angels also known va angels angel investor ...  
9   houston angel network provides capital earlyst...  
10  gobeyond enables individuals access angel inve...  
11  keiretsu forum northwest washingtonbased angel...  
12  new york angels professional angel organizatio...  
13  mumbai angels indian angel investment network ...  
14  indian angel network one leading angel investo...  
15  zillionize low friction startup investor portf...  
16  letsventure enables startups looking raise mon...  
17  anges québec network 230 angel investors inves...  
18  venturesouth early stage venture firm operates...  
19  boston harbor angels group proven business lea...

In [35]:
investors.to_csv('investors.csv')

# HEAT MAP FOR THE CORRELATION AMONG START-UPS

This algorithm will contribute to one of the app's functions, which is the "start-ups for you" or "investors that you might be interested in" section. 
To visualize the data better, we'll use the first 20 start-ups as an example to see their correlation and how the algorithm will work on our dataset.

In [34]:
dictionary = dict(zip(up['name'], up['new_des']))
dictionary

{'Campus Bubble': 'campus bubble cb academic community network linkedin professional community network facebook social community network cb provides academic institutions student powered cross platform private online community focused',
 'DueProps': 'unprecedented',
 'SalesLoft': 'build highquality prospect lists',
 'The Coca-Cola Company': 'cocacola journey digital magazine focuses important topics social causes news the cocacola company',
 'REscour': 'rescour data platform decision engine utilizes proprietary market analysis based massive data aggregation identify commercial real estate investment opportunities rescour onesizefitsall communicates custom tailored recommendations',
 'viaCycle': 'viacycle creates bicycle sharing technology flexible inexpensive easy use users unlock smart bikes instantly using cell phone operators place bikes anywhere kiosks special infrastructure required we enable bike',
 'Kabbage': 'kabbage delivers small businesses financing both ecommerce brickandmo

In [21]:
ids = list(dictionary.keys())

WHICH MEANS WE USE 269 DIFFERENT WORDS TO DESCRIBE 20 START-UPS